### PR TITLE
in setKit, check changed buildDirectory after instead of before refreshing expansions

### PR DESCRIPTION
**Pull request by david-fong**
### This changes visible behavior

The following changes are proposed:

- In `setKit`, call `_cleanPriorConfigurations` _after_ `_refreshExpansions`

## Other Notes/Information

This is just a draft / example. I have no idea if it is actually safe to move this check to go later. I also noticed similar code for `doSetConfigurePreset`, which might be a candidate for similar changes. What is `doSetConfigurePreset` for / for what user action does it get called?

### CODE REVIEW

**Comment by bobbrow** [comment]
Sorry for the delay on this.  We'll take a look at it next week.

**Comment by elahehrashedi** [comment]
@bobbrow I made a change to the PR.

File: src/drivers/cmakeDriver.ts [hunk 1]
```diff
@@ -514,14 +514,16 @@ export abstract class CMakeDriver implements vscode.Disposable {
...
-        const needs_clean = this.binaryDir === newBinaryDir && kitChangeNeedsClean(kit, this._kit);
-        await this.doSetKit(needs_clean, async () => {
+        const oldBinaryDir = this.binaryDir;
+        const needsCleanIfKitChange = kitChangeNeedsClean(kit, this._kit);
+        await this.doSetKit(async () => {
             await this._setKit(kit, preferredGenerators);
             await this._refreshExpansions();
+            const newBinaryDir = util.lightNormalizePath(await expand.expandString(this.config.buildDirectory, this.expansionOptions));
+            const needsCleanIfBuildDirectoryChange = (newBinaryDir !== oldBinaryDir);
+            if (needsCleanIfKitChange || needsCleanIfBuildDirectoryChange) {
```
Comment (bobbrow) [inline src/drivers/cmakeDriver.ts:524]: I don't think that changing the build directory is the right condition to delete the cache.  What I expected to see here was "if the kit changed, but the build directory didn't, then remove the prior cache" because I think it makes sense to expect that a different Kit is a different compiler and would configure the project differently (basically the same check that was happening on deleted line 521).  The rest of the fix seems correct because it defers the previous check until after the kit has been set and the `newBinaryDir` can be computed correctly.

**TL;DR:**
I think line 524 should be:
```ts
  if (needsCleanIfKitChange && newBinaryDir === oldBinaryDir) {
```

Comment (david-fong) [reply src/drivers/cmakeDriver.ts:524]: sorry about that. that was the old logic. I changed it in the first commit and I don't remember why. I think I was just confused.

Comment (david-fong) [reply src/drivers/cmakeDriver.ts]: just pushed a commit for this.

Comment (bobbrow) [reply src/drivers/cmakeDriver.ts]: No need to apologize. We appreciate you addressing this issue!

**Review by bobbrow — APPROVED** [review]

**Comment by bobbrow** [comment]
Something went (surprisingly) wrong with the tests after this change.  We'll need to investigate.

**Comment by bobbrow** [comment]
> Something went (surprisingly) wrong with the tests after this change. We'll need to investigate.

The good news is that it doesn't appear to be related to this change. I can repro it in `main`.  The bad news is that I don't have a root cause yet.

**Comment by david-fong** [comment]
The linux and windows runs don't say much, but the MacOS output shows something wrong that looks related to these changes (I don't know what exactly though). When you say "main", do you mean vscode-cmake-tools/main? because my PR branch is the main branch of my fork.

**Comment by david-fong** [comment]
Maybe it's an order of operations mistake. Before this PR, the call to `-`_cleanPriorConfiguration` was alwasy made _before_ calling the callback. But I moved it into the the callback and put it at the _end_ of the callback body, (now) _after_ these operations:

```
await this._setKit(kit, preferredGenerators);
await this._refreshExpansions();
```

**Comment by david-fong** [comment]
Oh but if that's really the issue, there's a chicken-egg problem: I moved it to go after those things because refreshing expansions is what enables getting the new path, which would be required to compare the new and old paths. In this case it might make sense to do more rework like I suggested in #2546 and make refreshExpansions more like a pure function.

**Comment by bobbrow** [comment]
I don't think there's anything wrong with your PR.  Other PR's are failing with the same issue (e.g. #2641).  A new version of VS Code was released this week and I suspect the test issues are related to that.  I'm going to commit this fix without the tests passing and work on the CI some more when I return from vacation.  With any luck the bug in VS Code will be addressed and fixed by then.  Otherwise, I'll report an issue when I return.

**Comment by david-fong** [comment]
Thanks! enjoy your vacation :)
